module/zfs/zfs_vfsops.c

```diff
@@ -1763,18 +1763,27 @@ zfs_remount(struct super_block *sb, int *flags, zfs_mnt_t *zm)
 {
 	zfsvfs_t *zfsvfs = sb->s_fs_info;
 	vfs_t *vfsp;
+	boolean_t issnap = dmu_objset_is_snapshot(zfsvfs->z_os);
 	int error;
 
 	error = zfsvfs_parse_options(zm->mnt_data, &vfsp);
 	if (error)
 		return (error);
 
-	zfs_unregister_callbacks(zfsvfs);
+	if ((issnap || !spa_writeable(dmu_objset_spa(zfsvfs->z_os))) &&
+	    !(*flags & MS_RDONLY)) {
+		*flags |= MS_RDONLY;
+		return (EROFS);
+	}
+
+	if (issnap)
+		zfs_unregister_callbacks(zfsvfs);
 	zfsvfs_vfs_free(zfsvfs->z_vfs);
 
 	vfsp->vfs_data = zfsvfs;
 	zfsvfs->z_vfs = vfsp;
-	(void) zfs_register_callbacks(vfsp);
+	if (issnap)
```

Comment:
You're right, the check is inverted. But we should be hitting `VERIFY(list_is_empty(list)) failed` from the dbu_evict thread in this case, so this means the test script is not covering every possible case.

I'll update this in a bit, thanks.